Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel $\rightarrow$ Restart) and then **run all cells** (in the menubar, select Cell $\rightarrow$ Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [1]:
NAME = "Akshata Padalkar"
COLLABORATORS = ""

---

Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel $\to$ Restart) and then **run all cells** (in the menubar, select Cell $\to$ Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [ ]:
NAME = ""
COLLABORATORS = ""

<div class="alert alert-block alert-info">
<h3>Student Information</h3> Please provide information about yourself.<br>
<b>Name</b>:Akshata Padalkar<br> 
<b>NetID</b>:anp178<br>
<b>Recitation #</b>:2<br>
<b>Notes to Grader</b> (optional):<br>
    I checked with TA Chetan, and he stated to ignore the error statements or comment the assert statements out for 1.4.c, 2.1, and 2.2. I have decided to ignore them and not comment out the assert statements. Additionally an announcement stated to use 30% of the data for testing. 
<br><br>
<b>IMPORTANT</b>
Your work will not be graded withour your initials below<br>
I certify that this lab represents my own work and I have read the RU academic intergrity policies at<br>
<a href="https://www.cs.rutgers.edu/academic-integrity/introduction">https://www.cs.rutgers.edu/academic-integrity/introduction </a><br>
<b>Initials</b>: AP (eg: NC for Naina Chaturvedi)

# CS 439 - Introduction to Data Science
### Spring 2025

## Lab 4 - Machine Learning (100 points)
### Due Date: April 15th 2025

In this lab, you will processing data for ML and building ML models. This notebook has multiple tasks that you need to complete. To complete a task, just double click on the cell (notebook must be opened in Jupyter/CodeBench) and enter your answer. Be sure to save the notebook often (ctrl-S or cmd-S) and execute the cell (Shift-Enter) to make sure you code is saved and executes correctly.

### Lab Discussions
Post any questions regarding this lab to canvas discussion board on Lab 03 or visit office hours.

## Course Policies

**Collaboration Policy**

Data science is a collaborative activity. While you may talk with others about the labs, we ask that you **write your solutions individually**. You are not allowed to copy code from other notebooks. Attempting to submit code that is not yours can result in penalties and academic disciplinary actions. Please consult course staff for any clarification regarding collaboration policies.

**AI Policy**

We **only allow** the use of responsible **AI tool youchat** to help you find information as you write code to solve labs. The use of AI should be done cautiously. It **does not** free you from understanding the code you write. Being able to read, interpret and test code written by AI is required for you to answer questions in quizzes and exams. **DO NOT use** other AI tools such as chatGPT and Bard.

### Running a Cell 

Try running the following cell.  If you are unfamiliar with Jupyter Notebooks, consider skimming [this tutorial](https://docs.jupyter.org/en/latest/) or selecting **Help** in the menu above. TO run the cell below, place the cursor on the cell and press **(SHIFT-ENTER)** or choose Run from the Jupyter menu above.

### NOTE: Whenever an interpretation is asked, you must provide one to receive full points.

For this lab we'll be using iris dataset found in seaborn's dataset repository

In [20]:
# import necessary libraries
import pandas as pd
import seaborn as sns
iris = sns.load_dataset('iris')

## Part 1: Preparing Data (60 Points)

### Task 1: Understanding the dataset (10 Points)
#### Task 1(a): Inspecting the data (5 Points)
- You'll be using the [New York City - East River Bicycle Crossings](https://www.kaggle.com/datasets/new-york-city/nyc-east-river-bicycle-crossings?resource=download) dataset
- Read the `nyc-east-river-bicycle-counts.csv` dataset into a pandas DataFrame and inspect it.

In [21]:
# Write your code here
file_name = 'cs439/Lab 04/nyc-east-river-bicycle-counts.csv'
def read_data(file_name):
    """
    Read the data from the file and return a Pandas DataFrame
    file_name: str, name of the file to read
    return: Pandas DataFrame
    """
    df = pd.read_csv(file_name)
    return df
    pass
 

# Do not modify the code below
try:
    data = read_data("nyc-east-river-bicycle-counts.csv")
    display(data)
    assert 'Date' in data.columns
    assert 'Day' in data.columns
    assert 'High Temp (°F)' in data.columns
    assert 'Low Temp (°F)' in data.columns
    assert 'Precipitation' in data.columns
    assert 'Brooklyn Bridge' in data.columns
    assert 'Manhattan Bridge' in data.columns
    assert 'Williamsburg Bridge' in data.columns
    assert 'Queensboro Bridge' in data.columns
    assert 'Total' in data.columns
    print("Your solution is correct!")
except Exception as e:
    print(f"OOPS! There is an error. {e}")

Unnamed: 0        Date         Day  High Temp (°F)  Low Temp (°F)  \
0             0  2016-04-01  2016-04-01            78.1           66.0   
1             1  2016-04-02  2016-04-02            55.0           48.9   
2             2  2016-04-03  2016-04-03            39.9           34.0   
3             3  2016-04-04  2016-04-04            44.1           33.1   
4             4  2016-04-05  2016-04-05            42.1           26.1   
..          ...         ...         ...             ...            ...   
198         205  2016-04-26  2016-04-26            60.1           46.9   
199         206  2016-04-27  2016-04-27            62.1           46.9   
200         207  2016-04-28  2016-04-28            57.9           48.0   
201         208  2016-04-29  2016-04-29            57.0           46.9   
202         209  2016-04-30  2016-04-30            64.0           48.0   

    Precipitation  Brooklyn Bridge  Manhattan Bridge  Williamsburg Bridge  \
0            0.01           1704.0              3126               4115.0   
1            0.15            827.0              1646               2565.0   
2            0.09            526.0              1232               1695.0   
3        0.47 (S)            521.0              1067               1440.0   
4               0           1416.0              2617               3081.0   
..            ...              ...               ...                  ...   
198          0.24           1997.0              3520               4559.0   
199             0           3343.0              5606               6577.0   
200             0           2486.0              4152               5336.0   
201          0.05           2375.0              4178               5053.0   
202             0           3199.0              4952               5675.0   

     Queensboro Bridge  Total  
0               2552.0  11497  
1               1884.0   6922  
2               1306.0   4759  
3               1307.0   4335  
4               2357.0   9471  
..                 ...    ...  
198             2929.0  13005  
199             4388.0  19914  
200             3657.0  15631  
201             3348.0  14954  
202             3606.0  17432  

[203 rows x 11 columns]

Your solution is correct!


In [22]:
# Inspect the data here
df = read_data("nyc-east-river-bicycle-counts.csv")
print(df.head())
print(df.info())       
print(df.describe())  

   Unnamed: 0        Date         Day  High Temp (°F)  Low Temp (°F)  \
0           0  2016-04-01  2016-04-01            78.1           66.0   
1           1  2016-04-02  2016-04-02            55.0           48.9   
2           2  2016-04-03  2016-04-03            39.9           34.0   
3           3  2016-04-04  2016-04-04            44.1           33.1   
4           4  2016-04-05  2016-04-05            42.1           26.1   

  Precipitation  Brooklyn Bridge  Manhattan Bridge  Williamsburg Bridge  \
0          0.01           1704.0              3126               4115.0   
1          0.15            827.0              1646               2565.0   
2          0.09            526.0              1232               1695.0   
3      0.47 (S)            521.0              1067               1440.0   
4             0           1416.0              2617               3081.0   

   Queensboro Bridge  Total  
0             2552.0  11497  
1             1884.0   6922  
2             1306.0   475

The data contains 203 rows and 11 columns.The columns include dates, weather info (temperature and precipitation), bicycle counts for four bridges, and a total count. Date and Day are stored as strings (object) and might have to be converted to datetime or other formats. Same with the Precipitation column it should most likely be converted to numeric. Some interesting data is that the High Temp (F) ranges from 39.9 F to 81 F, Low Temp (F) ranges from 26.1 F to 66 F, and the Total bicycle count per day ranges from 4,335 to 23,318. There is also no missing values in the data but possible inconsistencies in the data.

#### Task 1(b): Understand the Data (5 Points)

- Do you see a column named `Total`? What is its purpose?
- Visit the provided link and describe your understanding of the dataset, including a brief description of each column.

The Total column has the sum of the number of crossings for each bridge. Total = Brooklyn + Manhattan + Williamsburg + Queensboro. To describe each column briefly, we have a index column to keep track of entries, date column is the date of observation, day is a duplicate of the date column but most likely meant to be mapped to the days of the week, High Temp (°F) is the highest temperature recorded in NYC on that day (Fahrenheit), Low Temp (°F)	is the lowest temperature recorded in NYC on that day (Fahrenheit), Precipitation	is the amount of precipitation (in inches) which may include flags like (S) for snow. The Brooklyn Bridge column displays the bicycle count over the Brooklyn Bridge on that day, which is the same for the Manhattan Bridge, Williamsburg Bridge, and Queensboro Bridge. And as stated earlier the Total column displays the number of bicycle crossings across all 4 bridges on that day.

### Task 2: Preprocessing (10 Points)
- Do you see two columns that represent the same thing? If so, take appropriate action.
- Clean the data by removing null values, handling outliers, and addressing the redundant column.

In [23]:
# Write your code here
def pre_process(data):
    """
    Preprocess the data
    data: Pandas DataFrame
    return: Pandas DataFrame
    """
    if 'Unnamed: 0' in data.columns:
        data = data.drop(columns=['Unnamed: 0'])
        
    if 'Day' in data.columns:
        data = data.drop(columns=['Day'])

    data['Precipitation'] = (data['Precipitation'].str.replace(r"[^\d\.]", "", regex=True).replace('', '0').astype(float)                            )
    data = data.dropna()
    data['Manhattan Bridge'] = data['Manhattan Bridge'].astype(float)
    for col in ['Brooklyn Bridge' , 'Manhattan Bridge', 'Williamsburg Bridge', 'Queensboro Bridge', 'Total']:
        data = data[(data[col] >= 0) & (data[col] <= 50000)]
    return data

    pass

# Do not modify the code below
try:
    clean_data = pre_process(data)
    display(clean_data)
    assert 'Day' not in clean_data.columns
    assert 'Unnamed: 0' not in clean_data.columns
    assert 'Date' in clean_data.columns
    assert clean_data['Precipitation'].dtype == 'float64'
    print("Your solution is correct!")
except Exception as e:
    print(f"OOPS! There is an error. {e}")

Date  High Temp (°F)  Low Temp (°F)  Precipitation  \
0    2016-04-01            78.1           66.0           0.01   
1    2016-04-02            55.0           48.9           0.15   
2    2016-04-03            39.9           34.0           0.09   
3    2016-04-04            44.1           33.1           0.47   
4    2016-04-05            42.1           26.1           0.00   
..          ...             ...            ...            ...   
198  2016-04-26            60.1           46.9           0.24   
199  2016-04-27            62.1           46.9           0.00   
200  2016-04-28            57.9           48.0           0.00   
201  2016-04-29            57.0           46.9           0.05   
202  2016-04-30            64.0           48.0           0.00   

     Brooklyn Bridge  Manhattan Bridge  Williamsburg Bridge  \
0             1704.0            3126.0               4115.0   
1              827.0            1646.0               2565.0   
2              526.0            1232.0               1695.0   
3              521.0            1067.0               1440.0   
4             1416.0            2617.0               3081.0   
..               ...               ...                  ...   
198           1997.0            3520.0               4559.0   
199           3343.0            5606.0               6577.0   
200           2486.0            4152.0               5336.0   
201           2375.0            4178.0               5053.0   
202           3199.0            4952.0               5675.0   

     Queensboro Bridge  Total  
0               2552.0  11497  
1               1884.0   6922  
2               1306.0   4759  
3               1307.0   4335  
4               2357.0   9471  
..                 ...    ...  
198             2929.0  13005  
199             4388.0  19914  
200             3657.0  15631  
201             3348.0  14954  
202             3606.0  17432  

[203 rows x 9 columns]

Your solution is correct!


### Task 3: Feature Engineering (20 Points)
- Now that your data is clean, let's do some feature engineering.
- Extract the month and weekday (e.g., Monday, Tuesday, ...) from the date column, and create new columns 'Month' and 'Week'.

In [24]:
# Write your code here
def convert_date(data):
    """
    Convert the 'Date' column to datetime format (yyyy-mm-dd)
    data: Pandas DataFrame
    return: Pandas DataFrame
    """
    data['Date'] = pd.to_datetime(data['Date'].str.strip(), format='%Y-%m-%d')
    return data
    pass

def extract_month(data):
    """
    Extract the month from the 'Date' column and add a new column 'Month'
    data: Pandas DataFrame
    return: Pandas DataFrame
    """
    data['Month'] = data['Date'].dt.month
    return data
    pass

def exctract_weekday(data):
    """
    Extract the weekday from the 'Date' column and add a new column 'Weekday'
    data: Pandas DataFrame
    return: Pandas DataFrame
    """
    data['Weekday'] = data['Date'].dt.day_name()
    return data
    pass

def drop_date(data):
    """
    Drop the 'Date' column
    data: Pandas DataFrame
    return: Pandas DataFrame
    """
    data = data.drop(columns=['Date'])
    return data
    pass

# Do not modify the code below
try:
    feature_data = convert_date(clean_data)
    feature_data = extract_month(feature_data)
    feature_data = exctract_weekday(feature_data)
    feature_data = drop_date(feature_data)
    display(feature_data)
    assert 'Month' in feature_data.columns
    assert 'Weekday' in feature_data.columns
    assert 'Date' not in feature_data.columns
    print("Your solution is correct!")
except Exception as e:    
    print(f"OOPS! There is an error. {e}")

High Temp (°F)  Low Temp (°F)  Precipitation  Brooklyn Bridge  \
0              78.1           66.0           0.01           1704.0   
1              55.0           48.9           0.15            827.0   
2              39.9           34.0           0.09            526.0   
3              44.1           33.1           0.47            521.0   
4              42.1           26.1           0.00           1416.0   
..              ...            ...            ...              ...   
198            60.1           46.9           0.24           1997.0   
199            62.1           46.9           0.00           3343.0   
200            57.9           48.0           0.00           2486.0   
201            57.0           46.9           0.05           2375.0   
202            64.0           48.0           0.00           3199.0   

     Manhattan Bridge  Williamsburg Bridge  Queensboro Bridge  Total  Month  \
0              3126.0               4115.0             2552.0  11497      4   
1              1646.0               2565.0             1884.0   6922      4   
2              1232.0               1695.0             1306.0   4759      4   
3              1067.0               1440.0             1307.0   4335      4   
4              2617.0               3081.0             2357.0   9471      4   
..                ...                  ...                ...    ...    ...   
198            3520.0               4559.0             2929.0  13005      4   
199            5606.0               6577.0             4388.0  19914      4   
200            4152.0               5336.0             3657.0  15631      4   
201            4178.0               5053.0             3348.0  14954      4   
202            4952.0               5675.0             3606.0  17432      4   

       Weekday  
0       Friday  
1     Saturday  
2       Sunday  
3       Monday  
4      Tuesday  
..         ...  
198    Tuesday  
199  Wednesday  
200   Thursday  
201     Friday  
202   Saturday  

[203 rows x 10 columns]

Your solution is correct!


### Checkpoint!
- Use the cleaned dataset `nyc-east-river-bicycle-counts-cleaned.csv` for the following tasks.

In [25]:
data = pd.read_csv("nyc-east-river-bicycle-counts-cleaned.csv")
print(data.head())

   High Temp (°F)  Low Temp (°F)  Precipitation  Brooklyn Bridge  \
0            78.1           66.0           0.01           1704.0   
1            55.0           48.9           0.15            827.0   
2            39.9           34.0           0.09            526.0   
3            44.1           33.1           0.47            521.0   
4            42.1           26.1           0.00           1416.0   

   Manhattan Bridge  Williamsburg Bridge  Queensboro Bridge  Total  Month  \
0              3126               4115.0             2552.0  11497      4   
1              1646               2565.0             1884.0   6922      4   
2              1232               1695.0             1306.0   4759      4   
3              1067               1440.0             1307.0   4335      4   
4              2617               3081.0             2357.0   9471      4   

    Weekday  
0    Friday  
1  Saturday  
2    Sunday  
3    Monday  
4   Tuesday  


### Task 4: Encoding (20 Points)

- You've learned a lot about data! Now let's teach the computer about the data. Yeah—machine learning!
- Before getting into it, remember that the computer only understands numbers, so encode a few columns into numbers.

#### Task 4(a): Identify the columns to encode (5 Points)
- You obviously know that Weekday should be encoded, but there is one more column that should be encoded. Can you spot it?
- It is a column where its numerical values represent something.

Wtite your answer here (you can clear this text)
Announcement on April 2 said to ignore this question. 

#### Task 4(b): Identify the encoding technique (5 Points)
- Now that you've found the columns to encode, which encoding technique would be appropriate here, one-hot encoding or label encoding?

Since label encoding is used for ordinal data and one-hot encoding is used for nominal data, for the weekday column since no day has superiority I believe we should use one-hot encoding. 

#### Task 4(c): Implementing Encoding (10 Points)
- Now that you've identified the columns and the appropriate encoding technique, apply the encoding to the dataset using pandas or scikit-learn.

In [26]:
# Import necessary libraries
def encode_column(data):
    """
    Encode the 'Weekday' and the 'other column' using the appropriate encoding technique.
    data: Pandas DataFrame
    return: Pandas DataFrame with the encoded columns, and the original columns removed
    """
    data = pd.get_dummies(data, columns=['Weekday'], drop_first= False)
    return data
    pass
try:
    encoded_data = encode_column(data)
    display(encoded_data.shape)
    
    if encoded_data.shape[1] == 16:
        assert False, "You might have forgotten to drop the original columns"
    assert encoded_data.shape[1] == 14
    
    print("Your solution might be correct!")
except Exception as e:
    print(f"OOPS! There is an error. {e}")

(203, 16)

OOPS! There is an error. You might have forgotten to drop the original columns


## Part 2: Model Building (40 Points)

### Task 1: Split Train and Test Data (10 Points)

- Split the dataset into training and testing sets, with 20% of the data in the testing set (use random_state=42).

In [27]:
# Import necessary libraries
from sklearn.model_selection import train_test_split
#annoucement states that split is 30% and 70%
def split_data(data):
    """
    Split the data into features and target
    data: Pandas DataFrame
    return: X_train, X_test, y_train, y_test
    """
    X = data.drop(columns=['Total'])
    y = data['Total']
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
    return X_train, X_test, y_train, y_test
    pass
# Do not modify the code below
try:
    X_train, X_test, y_train, y_test = split_data(encoded_data)
    assert X_train.shape == (142, 13)
    assert X_test.shape == (61, 13)
    assert y_train.shape == (142,)
    assert y_test.shape == (61,)
    print("Your solution is correct!")
except Exception as e:
    print(f"OOPS! There is an error. {e}")


OOPS! There is an error. 


### Task 2: Standardize the data (10 Points)

- Let's standardize our data to ensure that the coefficients can be meaningfully compared.
- Make sure you fit the standardizer on the training set and use it for both training and testing sets. Then, explain why this is necessary (think about what the test data represents).

In [28]:
# Import necessary libraries
from sklearn.preprocessing import StandardScaler

def scale_data(X_train, X_test):
    """
    Scale the features using StandardScaler
    X_train: Pandas DataFrame
    X_test: Pandas DataFrame
    return: X_train_scaled, X_test_scaled
    """
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)
    
    return X_train_scaled, X_test_scaled
    pass

# Do not modify the code below
try:
    X_train_scaled, X_test_scaled = scale_data(X_train, X_test)
    assert X_train_scaled.shape == (142, 13)
    assert X_test_scaled.shape == (61, 13)
    print("Your solution might be correct!")
except Exception as e:
    print(f"OOPS! There is an error. {e}")


OOPS! There is an error. 


If we fit the scaler on the entire dataset (train + test), we let the model look into the test data, which would not be good for realistic and true testing since we want to test on unseen data. By fitting on the training set and transforming both sets, we ensure the same scaling logic is applied everywhere. This way we avoid having different means and std devs skewing the inputs during evaluation.

### Task 3: Model Building (10 Points)
- Build a multiple linear regression model (5 Points)
- Build a support vector regression model using the `sigmoid` kernel (5 Points)

In [30]:
# Import necessary libraries
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVR

def train_linear_model(X_train, y_train):
    """
    Train a Linear Regression model
    X_train: Numpy Array
    y_train: Numpy Array
    return: Trained Linear Regression model
    """
    lr_model = LinearRegression()
    lr_model.fit(X_train, y_train)
    return lr_model
    pass

def train_svm_model(X_train, y_train):
    """
    Train a Support Vector Machine model
    X_train: Numpy Array
    y_train: Numpy Array
    return: Trained Support Vector Machine model
    """
    svr_model = SVR(kernel='sigmoid')
    svr_model.fit(X_train, y_train)
    return svr_model
    pass

# Do not modify the code below
try:
    from sklearn.linear_model import LinearRegression
    from sklearn.svm import SVR
    lr_model = train_linear_model(X_train_scaled, y_train)
    svr_model = train_svm_model(X_train_scaled, y_train)
    assert isinstance(lr_model, LinearRegression)
    assert isinstance(svr_model, SVR)
    print("Your solution might be correct!")
except Exception as e:
    print(f"OOPS! There is an error. {e}")

Your solution might be correct!


### Task 4: Model Evaluation (10 Points)

- Evaluate your models' performance.

In [31]:
# Import necessary libraries
from sklearn.metrics import r2_score, mean_squared_error
import numpy as np
# Write your code here
def evaluate_model(model, X_test, y_test):
    y_pred = model.predict(X_test)
    r2 = r2_score(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    return {"r2 Score": r2, 
            "MSE": mse,}
    
results1 = evaluate_model(lr_model, X_test_scaled, y_test)
results2 = evaluate_model(svr_model, X_test_scaled, y_test)
for metric, value in results1.items():
    print("Linear Model ",f"{metric}: {value:.4f}")
for metric, value in results2.items():
    print("SVR Model ",f"{metric}: {value:.4f}")

Linear Model  r2 Score: 1.0000
Linear Model  MSE: 0.0000
SVR Model  r2 Score: -0.1040
SVR Model  MSE: 34833699.5383


##### Explain why you chose the metric you used, and write your interpretation on why one of the models is better than the other.

I decided to use R2 and MSE to interpret the models because R2 tells us how well the model explains the data, while MSE shows how far off the predictions are on average. Using both helps to see if a model is accurate and reliable for making predictions. Knowing this and seeing the results of the metric calculations, the Linear model is clearly better, with a high positive score of 1 for the R2 metric which means it fits the training data well. It also has an MSE of 0 which means basically no prediction error on the test set. The SVR model on the other hand has a negative R2 score, which means it does not fit the training data well at all, and a very large MSE value which means high prediction error, possibly using a different kernal may help this situation, but for now as it is, the linear model is better. 

<div class="alert alert-block alert-info">
<h2>Submission Instructions</h2> 
<b> File Name:</b> Please rename the file as netID_Lab_02_WebScapping_and_EDA (eg: netID = nc123)<br>
<b> Submit through: </b> CodeBench <br>
<b>Warning:</b> Failure to follow directions may result in loss points.<br>
</div>